### PR TITLE
SignalFx exporter: Add host metadata synchronization

### PR DESCRIPTION
**Pull request by dmitryax**
**Description:** 
This change introduces host metadata updates functionality in Signalfx exporter. Synchronization happens once first metric batch is received. It looks for dimensions set by resourcedetection processor

### CODE REVIEW

**Comment by dmitryax** [comment]
as discussed with @jrcamp it will be changes to use cloud unique id instead of host

**Comment by jrcamp** [comment]
@dmitryax I think the PR has become too big. I think it'd be good to break it up at least into two pieces, one being stuff that came from smart agent that's already been vetted there and the second is new code that may need a closer look.

**Comment by dmitryax** [comment]
@jrcamp broke it down as you suggested. The first part: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1118

File: exporter/signalfxexporter/hostmetadata/metadata.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		s.synced = true
+		return
+	}
+
+	metadataUpdate := s.prepareMetadataUpdate(props, hostID)
+	err := s.dimClient.PushMetadata([]*collection.MetadataUpdate{metadataUpdate})
+	if err != nil {
+		s.logger.Error("Failed to push host metadata update", zap.Error(err))
+		return
+	}
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go:82]: thanks. i'll do that using sync.Once

File: exporter/signalfxexporter/hostmetadata/metadata.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	extraHostIDAttribute = "host"
+)
+
+// Syncer is a config structure for host metadata syncer.
+type Syncer struct {
+	logger    *zap.Logger
+	dimClient dimensions.MetadataUpdateClient
+	synced    bool
+
+	done chan struct{}
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go:36]: not anymore. good catch, thanks.

File: exporter/signalfxexporter/hostmetadata/metadata.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/signalfxexporter/dimensions"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sclusterreceiver/collection"
+)
+
+var (
+	// Resource attribute that used as fallback if there are no cloud provider attributes.
+	extraHostIDAttribute = "host"
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go:27]: that makes sense. I'll update it in this PR

[the remark this answers is not in the document]
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go]: updated

File: exporter/signalfxexporter/hostmetadata/metadata.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go]: fixed

File: exporter/signalfxexporter/go.mod [hunk 4]
```diff
@@ -15,6 +15,7 @@ require (
 	go.uber.org/multierr v1.5.0
 	go.uber.org/zap v1.16.0
 	google.golang.org/protobuf v1.25.0
+	gotest.tools v2.2.0+incompatible
```
Comment (jrcamp) [inline exporter/signalfxexporter/go.mod:18]: Use testify assert instead of adding new dep.

**Comment by jrcamp** [comment]
lgtm after the assert thing is fixed and #1100 is merged (should be soon) so that can be removed here.

File: exporter/signalfxexporter/go.mod (hunk 4 above)
Comment (dmitryax) [reply exporter/signalfxexporter/go.mod:18]: oh nice catch! I got it automatically imported

Comment (jrcamp) [reply exporter/signalfxexporter/go.mod]: Yeah it's always importing the wrong one for me haha. Major pain.

**Comment by dmitryax** [comment]
@jrcamp , PR rebased

**Review by jrcamp — APPROVED** [review]

**Pull request merged**